core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java

```diff
@@ -0,0 +1,154 @@
+        tasks = new TaskInfo[in.readVInt()];
+        for (int i = 0; i < tasks.length; i++) {
+            tasks[i] = new TaskInfo(in);
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeVInt(tasks.length);
+        for (TaskInfo task : tasks) {
+            task.writeTo(out);
+        }
+    }
+
+    /**
+     * Returns the list of tasks by node
+     */
+    public Map<String, List<TaskInfo>> getNodes() {
+        if (nodes != null) {
+            return nodes;
+        }
+        Map<String, List<TaskInfo>> nodeTasks = new HashMap<>();
+
+        Set<String> nodes = new HashSet<>();
+        for (TaskInfo shard : tasks) {
+            nodes.add(shard.node());
+        }
+
+        for (String node : nodes) {
+            List<TaskInfo> tasks = new ArrayList<>();
+            for (TaskInfo taskInfo : this.tasks) {
+                if (taskInfo.node().equals(node)) {
+                    tasks.add(taskInfo);
+                }
+            }
+            nodeTasks.put(node, tasks);
+        }
+        this.nodes = nodeTasks;
+        return nodeTasks;
+    }
+
+    public TaskInfo[] getTasks() {
+        return tasks;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        if (getTaskFailures() != null && getTaskFailures().length > 0) {
+            builder.startArray(Fields.TASK_FAILURES);
+            for (TaskOperationFailedException ex : getTaskFailures()){
+                builder.value(ex);
+            }
+            builder.endArray();
+        }
+
+        if (getNodeFailures() != null && getNodeFailures().length > 0) {
+            builder.startArray(Fields.NODE_FAILURES);
+            for (FailedNodeException ex : getNodeFailures()){
+                builder.value(ex);
+            }
+            builder.endArray();
+        }
+
+        builder.startObject(Fields.NODES);
+        for (Map.Entry<String, List<TaskInfo>> entry : getNodes().entrySet()) {
+            String node = entry.getKey();
+            builder.startObject(node, XContentBuilder.FieldCaseConversion.NONE);
+            builder.startArray(Fields.TASKS);
+            for(TaskInfo task : entry.getValue()) {
+                task.toXContent(builder, params);
+            }
+            builder.endArray();
+            builder.endObject();
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    static final class Fields {
+        static final XContentBuilderString NODES = new XContentBuilderString("nodes");
+        static final XContentBuilderString TASKS = new XContentBuilderString("tasks");
+        static final XContentBuilderString TASK_FAILURES = new XContentBuilderString("task_falures");
+        static final XContentBuilderString NODE_FAILURES = new XContentBuilderString("node_falures");
+    }
+
+    @Override
+    public String toString() {
```

Comment:
maybe `BaseTasksResponse` can extend `ToXContentToBytes` then we get these toString methods for free?